File: test/generate_fixtures_bootstrap.R
#!/usr/bin/env Rscript
#
# Level-1 fixture generator (R / ground-truth side) — bootstrap path.
#
# Reads the HDF5 fixture written by `generate_fixtures_bootstrap.jl`, applies
# `cAIC4`'s `conditionalBootstrap` bias-correction arithmetic to each case, and writes
# the reference `rho_ref` back into the same file. This is the cAIC4 ground truth the
# Level-1 test compares against (rtol = 1e-6, atol = 1e-10) — the shared-input fixture
# the bootstrap path was missing.
#
# Why we isolate the arithmetic from the function. `conditionalBootstrap(object, B)`
# is a closed pipeline: it `simulate()`s, `refit()`s, and only then applies the
# bias-correction formula on lines 23–25 of `R/conditionalBootstrap.R`. The Level-1
# unit needs *just* that arithmetic, on fixed `Y*` / `Ŷ*` matrices, so we hard-code
# the formula here and `source()` the cAIC4 file purely to **pin the formula** via a
# textual self-check (`grepl` on the function body): if a `cAIC4` bump silently moves
# either the row-mean centring or the `(B−1)` divisor, fixture generation stops loud.
# The committed cAIC4 v1.1 lines this matches are reproduced verbatim in the comment
# above `bootBC_arithmetic` below.
#
# HDF5 reader: `rhdf5` — same as `generate_fixtures.R`; the ADR-0003 addendum
# (2026-05-27) covers the choice.
#
# Env vars:
#   CAIC4_SRC  path to the cAIC4 source tree (default /private/tmp/cAIC4_src)
#   FIXTURE    path to the HDF5 fixture (default <script dir>/fixtures/bootstrap_level1.h5)
#
# Usage:  Rscript test/generate_fixtures_bootstrap.R

suppressMessages(library(rhdf5))

caic4_src <- Sys.getenv("CAIC4_SRC", "/private/tmp/cAIC4_src")
fixture <- Sys.getenv("FIXTURE", "")
if (!nzchar(fixture)) {
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- sub("^--file=", "", args[grep("^--file=", args)])
  here <- if (length(file_arg)) dirname(normalizePath(file_arg)) else "test"
  fixture <- file.path(here, "fixtures", "bootstrap_level1.h5")
}

cb_src <- file.path(caic4_src, "R", "conditionalBootstrap.R")
stopifnot("cAIC4 source not found; set CAIC4_SRC" = file.exists(cb_src))
source(cb_src)
caic4_version <- tryCatch(
  unname(read.dcf(file.path(caic4_src, "DESCRIPTION"))[, "Version"]),
  error = function(e) "unknown"
)

# `cAIC4` v1.1 `R/conditionalBootstrap.R` lines 23–25 (bias-correction arithmetic):
#   dataMatrix <- dataMatrix - rowMeans(dataMatrix)
#   bootBC     <- sum(workingEta * dataMatrix) /
#                 ((BootStrRep - 1) * sigma(object)^2)
# Isolated below as a pure function of `(Ystar, Yhatstar, sigma)`; B = ncol(Ystar).
bootBC_arithmetic <- function(Ystar, Yhatstar, sigma) {
  centered <- Ystar - rowMeans(Ystar)
  B <- ncol(Ystar)
  sum(Yhatstar * centered) / ((B - 1) * sigma^2)
}

# Pin the cAIC4 formula textually: any silent drift in either the centring or the
# divisor must fail fixture regeneration loudly (CLAUDE §10 — don't paper over). The
# deparsed body is whitespace-normalised before matching, because R's `deparse` may
# insert a line break inside `(BootStrRep - 1) * sigma(object)^2`.
body_str <- gsub("\\s+", "", paste(deparse(body(conditionalBootstrap)), collapse = ""))
stopifnot(
  "cAIC4 conditionalBootstrap row-mean centring drifted" =
    grepl("dataMatrix-rowMeans(dataMatrix)", body_str, fixed = TRUE),
  "cAIC4 conditionalBootstrap (B-1)*sigma^2 divisor drifted" =
    grepl("sum(workingEta*dataMatrix)/((BootStrRep-1)*sigma(object)^2)",
          body_str, fixed = TRUE)
)

stopifnot("fixture not found; run generate_fixtures_bootstrap.jl first" =
            file.exists(fixture))
on.exit(h5closeAll())

put <- function(path, value) { # idempotent dataset write
  try(h5delete(fixture, path), silent = TRUE)
  h5write(value, fixture, path)
}

top <- h5ls(fixture, recursive = FALSE)
case_names <- setdiff(top$name[top$otype == "H5I_GROUP"], "meta")

for (name in case_names) {
  d <- h5read(fixture, name)
  Ystar <- d$Ystar
  Yhatstar <- d$Yhatstar
  sigma <- as.numeric(d$sigma)
  rho <- bootBC_arithmetic(Ystar, Yhatstar, sigma)
  put(paste0(name, "/rho_ref"), as.numeric(rho))
  cat(sprintf("  %-10s rho_ref = %.12g  (n=%d, B=%d)\n",
              name, rho, nrow(Ystar), ncol(Ystar)))
}

put("meta/cAIC4_version", caic4_version)
put("meta/rhdf5_version", as.character(packageVersion("rhdf5")))
put("meta/R_version", R.version.string)

cat(sprintf("Wrote rho_ref for %d case(s) to %s (cAIC4 %s).\n",
            length(case_names), fixture, caic4_version))

File: test/generate_fixtures_bootstrap.jl
#!/usr/bin/env julia
#
# Level-1 fixture generator (Julia side) — bootstrap path / shared-input parity gate.
#
# Writes seeded, synthetic `yhat`, `sigma`, `Y*`, `Ŷ*` matrices into an HDF5 file. The
# R side (`generate_fixtures_bootstrap.R`) reads it, evaluates `cAIC4`'s
# `conditionalBootstrap` bias-correction arithmetic on each case, and writes the
# reference `rho_ref` back. CI reads the resulting fixture in Julia with **no R** and
# compares against `ConditionalAIC.DofLMM.efron_penalty` at the Level-1 tolerance
# (rtol = 1e-6, atol = 1e-10).
#
# Why the inputs are synthetic, not from a real bootstrap: `conditionalBootstrap`'s
# bias-correction arithmetic is `bootBC = Σ ŷ*·(y*−ȳ*) / ((B−1)σ²)` — pure linear
# algebra in `Y*`, `Ŷ*`, `σ`. The simulate-and-refit machinery only *produces* `Y*`
# and `Ŷ*`; the Level-1 isolation tests the arithmetic itself on **fixed** matrices,
# independent of which optimiser, RNG, or family produced them. This is the same
# isolation principle as `generate_fixtures.jl` for `calculateGaussianBc`.
#
# Usage: `julia --project=test test/generate_fixtures_bootstrap.jl`

using HDF5
using LinearAlgebra
using Random
using Statistics: mean

const FIXTURE = joinpath(@__DIR__, "fixtures", "bootstrap_level1.h5")

"""
    bootstrap_case(rng; n, B, sigma, mean_scale = 1.0, signal = 0.5)

Build one synthetic `(yhat, sigma, Y*, Ŷ*)` case. `Y*` is `ŷ + σε` with
`ε ~ N(0, Iₙ)` (mimicking `lme4`'s `simulate(..., use.u = TRUE)` parametric draw),
and `Ŷ*` is correlated with `Y*` via `signal` to give a non-trivial covariance
penalty — exactly the structure a real `predict(refit(..., newresp = y*))` produces.
The arithmetic is parametrisation-neutral, so these matrices are valid inputs for
`conditionalBootstrap`'s bias-correction formula regardless of how they were generated.
"""
function bootstrap_case(
    rng::AbstractRNG;
    n::Int,
    B::Int,
    sigma::Float64,
    mean_scale::Float64=1.0,
    signal::Float64=0.5,
)
    B >= 2 || throw(ArgumentError("B must be ≥ 2 (cAIC4's (B−1) divisor); got $B"))
    yhat = mean_scale .* randn(rng, n)
    epsY = randn(rng, n, B)
    Ystar = yhat .+ sigma .* epsY
    epsH = randn(rng, n, B)
    # Yhatstar correlated with Y* (a `predict(refit(...))` would couple them similarly).
    Yhatstar = yhat .+ signal .* (Ystar .- yhat) .+ sqrt(1 - signal^2) .* sigma .* epsH
    return (; yhat, sigma, Ystar, Yhatstar)
end

"""
    build_bootstrap_cases() -> Dict{String, NamedTuple}

The seeded Level-1 cases for the bootstrap shared-input fixture. Spans `B = 2`
(the minimum the `(B−1)` divisor admits), small/mid/large `n` and `B`, and a high-
signal case where `ŷ*` is strongly correlated with `y*`.
"""
function build_bootstrap_cases()
    cases = Dict{String,NamedTuple}()

    # tracer: n=8, B=20 — moderate signal, σ=1.5
    cases["tracer"] = bootstrap_case(
        MersenneTwister(0x424f_4f54);            # "BOOT"
        n=8,
        B=20,
        sigma=1.5,
        mean_scale=2.0,
        signal=0.5,
    )

    # mid: n=25, B=100, weaker signal, smaller σ
    cases["mid"] = bootstrap_case(
        MersenneTwister(0x4d_4944_3030);         # "MID00"
        n=25,
        B=100,
        sigma=0.8,
        mean_scale=1.0,
        signal=0.3,
    )

    # large: n=50, B=500 — exercises the larger Σ; strong signal
    cases["large"] = bootstrap_case(
        MersenneTwister(0x4c_4152_4745);         # "LARGE"
        n=50,
        B=500,
        sigma=2.5,
        mean_scale=3.0,
        signal=0.7,
    )

    # min_B: B = 2 (the smallest value admitting cAIC4's (B−1) divisor)
    cases["min_B"] = bootstrap_case(
        MersenneTwister(0x4d_494e_5f42);         # "MIN_B"
        n=6,
        B=2,
        sigma=1.0,
        mean_scale=0.5,
        signal=0.4,
    )

    return cases
end

function write_bootstrap_fixture(path, cases)
    mkpath(dirname(path))
    h5open(path, "w") do f
        for (name, c) in cases
            g = create_group(f, name)
            g["yhat"] = c.yhat
            g["sigma"] = c.sigma
            g["Ystar"] = c.Ystar
            g["Yhatstar"] = c.Yhatstar
            g["n"] = size(c.Ystar, 1)
            g["B"] = size(c.Ystar, 2)
        end
        meta = create_group(f, "meta")
        meta["generator"] = "ConditionalAIC.jl test/generate_fixtures_bootstrap.jl"
        meta["julia_version"] = string(VERSION)
        meta["hdf5_jll_version"] = string(HDF5.API.h5_get_libversion())
    end
    return path
end

if abspath(PROGRAM_FILE) == @__FILE__
    cases = build_bootstrap_cases()
    write_bootstrap_fixture(FIXTURE, cases)
    @info "Wrote bootstrap Level-1 fixture" path = FIXTURE ncases = length(cases) cases = sort(
        collect(keys(cases))
    )
end
